hpx/runtime/actions/continuation.hpp

```diff
@@ -286,34 +286,123 @@ namespace hpx { namespace actions
+        {
+            try {
+                util::invoke(std::forward<F>(f), std::forward<Ts>(vs)...);
+                cont->trigger();
+            }
+            catch (...) {
+                // make sure hpx::exceptions are propagated back to the client
+                cont->trigger_error(boost::current_exception());
+            }
+        }
+
+        // special handling of actions returning a future
+        template <typename Result, typename RemoteResult, typename Future>
+        void deferred_trigger(std::false_type, std::unique_ptr<continuation> cont, Future result)
+        {
+            try {
+                HPX_ASSERT(result.is_ready());
+                static_cast<typed_continuation<Result, RemoteResult>*>(cont.get())->
+                    trigger(result.get());
+            }
+            catch (...) {
+                // make sure hpx::exceptions are propagated back to the client
+                cont->trigger_error(boost::current_exception());
+            }
+        }
+
+        template <typename Result, typename RemoteResult, typename Future>
+        void deferred_trigger(std::true_type, std::unique_ptr<continuation> cont, Future result)
+        {
+            try {
+                HPX_ASSERT(result.is_ready());
+                cont->trigger();
+            }
+            catch (...) {
+                // make sure hpx::exceptions are propagated back to the client
+                cont->trigger_error(boost::current_exception());
+            }
+        }
+
+        template <typename Future, typename F, typename ...Ts>
+        void trigger_impl_future(boost::mpl::true_,
+            std::unique_ptr<continuation> cont, F&& f, Ts&&... vs)
+        {
+            typedef
+                typename traits::future_traits<
+                    Future
+                >::type type;
+            typedef
+                typename detail::remote_action_result<type>::type
+                remote_result_type;
+            typedef typename std::is_void<type>::type is_void;
+
+            Future result = util::invoke(std::forward<F>(f),
+                    std::forward<Ts>(vs)...);
+
+            if(result.is_ready())
+            {
+                detail::deferred_trigger<type, remote_result_type>(
+                    is_void(), std::move(cont), std::move(result));
+                return;
+            }
+
+            void (*fun)(is_void, std::unique_ptr<continuation>, Future)
+                = detail::deferred_trigger<type, remote_result_type, Future>;
+
+            result.then(
+                hpx::util::bind(
+                    hpx::util::one_shot(fun)
+                  , is_void()
+                  , std::move(cont)
+                  , util::placeholders::_1
+                )
+            );
+        }
+
+        template <typename R, typename F, typename ...Ts>
+        void trigger_impl_future(boost::mpl::false_,
```

Comment:
Same as above. Please use only one `boost::mpl::bool_` or `std::integral_constant`. The rest of the code now uses the latter...